### DISCUSSION TITLE
Clarification on Sliding Window Inference

### DISCUSSION THREAD

**Discussion by sh3rlock14** [Project-MONAI/MONAI#5701] [Q&A] [answered]
Hello everybody!
I am not grasping entirely how `sliding_window_inference()` works, in particular:

1. I guess that by setting `overlap=0` and `roi_size=input_size`, it would be as using SimpleInferer, right?
2. in a validation step (as in pytorch lightning) I would call directly the model on my input *e.g.* `output = model(input)`, it is the same as calling `output = inferer(inputs, model)`, with `inferer=SimpleInferer`, right?
3. what does `sw_batch_size` param refers to? 
I followed the tutorial about the [post_transform](https://github.com/Project-MONAI/MONAIBootcamp2020/blob/master/day1notebooks/lab3_post_transforms.ipynb) but I could not figure it out what was happening when changing its value from `1` to higher values. 

For the experiment I chose `overlap=0` so that I could see clearly what was happening at each window, and then choose the `roi_size` to be equal to input size.

```
class ToyModel:
    # A simple model generates the output by adding an integer `pred` to input.
    # each call of this instance increases the integer by 1.
    pred = 0
    def __call__(self, input):
        self.pred = self.pred + 1
        return input + self.pred
```

```
input_tensor = torch.zeros(3, 1, 200, 200) # BCWH
output_tensor = sliding_window_inference(
    inputs=input_tensor, 
    predictor=ToyModel(), 
    roi_size=(200, 200), # Infer on the whole image at once
    sw_batch_size=2, 
    overlap=0., 
    mode="constant")
```

`px.imshow(output_tensor[0, 0])`
 
![newplot(2)](https://user-images.githubusercontent.com/28751714/206925962-afe7845d-3f62-4ce4-b992-b1c18f0a6761.png)

`px.imshow(output_tensor[1, 0])`

![newplot(2)](https://user-images.githubusercontent.com/28751714/206925962-afe7845d-3f62-4ce4-b992-b1c18f0a6761.png)

`px.imshow(output_tensor[2, 0])`

![newplot(3)](https://user-images.githubusercontent.com/28751714/206926088-f77f614f-77fb-41ad-8569-290367270e0b.png)

**notice**: the value of the preds in the third image is `2`, so given my `batch_size` and `sw_batch_size`, what has happened is that the first 2 inputs were considered, the model ran the inference on them (first step) and then the last input was considered and the model ran the inference on it (second step)... is it so?

Basically with `roi_size` I decide "how big" the patches taken from a single image are, while with "sw_batch_size" I decide "how many images to consider" while doing inference within a single step: can you confirm?

My doubt arises from the fact that when `roi_size` < `input_size`, the behaviour I just described seems to be applied not on the "batch dim",  but rather on the "window dim", *i.e.:*

```
input_tensor = torch.zeros(3, 1, 200, 200)
output_tensor = sliding_window_inference(
    inputs=input_tensor, 
    predictor=ToyModel(), 
    roi_size=(50, 50), 
    sw_batch_size=2, 
    overlap=0., 
    mode="constant")
```

![newplot(4)](https://user-images.githubusercontent.com/28751714/206928240-dfb024be-ef6c-4a33-84fc-5e992647fbe2.png)

so, in the batch there are not windows from different images, but rather subsequent windows from the same input.
Is it so? Is it working this way because of memory efficiency?

I can see that this reasoning can be generalized to the one I made before, where the subwindows match exactly the original input and the batches become "traditional batches".

**Comment by KumoLiu** [comment, marked as answer]
Hi @sh3rlock14, when we use `SlidingWindowInferer`, we mainly consider when the original data is too large and the memory is not enough, at this time we can randomly crop during training and predict by sliding window during inference.
For `sw_batch_size`, you can simply treat it as batch dimension during inference, the total num of windows is `windows_num_per_image * batch_size`.  Additionally, if your `roi_size` is equal to `input_size` you can use `SimpleInferer` directly instead of `SlidingWindowInferer`. For the `overlap` arg, it is mainly to ensure more accurate results of the boundary when doing sliding windows.
https://github.com/Project-MONAI/MONAI/blob/4f584a945a5deb5622fb09d165fe13f5cf46e3d7/monai/inferers/utils.py#L147-L149

Hope it can help you, thanks!
